### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by DimitrisPapac**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+[workspace]
+members = [
+  "curves/bn254",
+  "models",
+  "test-utils",
+]
+resolver = "2"
+
+[workspace.package]
+version = "0.4.2"
```
Comment (lgiussan) [inline Cargo.toml]: Why `0.4.2` specifically?

File: Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+authors = ["Horizen Labs <admin@horizenlabs.io>"]
+repository = "https://github.com/DimitrisPapac/accelerated-bn-cryptography.git"   # "https://github.com/zkVerify/accelerated-bn-cryptography"
```
Comment (lgiussan) [inline Cargo.toml]: Please update to `https://github.com/zkVerify/accelerated-bn-cryptography`

File: curves/bn254/src/curves/mod.rs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+impl<H: CurveHooks> BnConfig for Config<H> {
+    const X: &'static [u64] = <ArkConfig as ArkBnConfig>::X;
+    const X_IS_NEGATIVE: bool = <ArkConfig as ArkBnConfig>::X_IS_NEGATIVE;
+    const TWIST_TYPE: TwistType = <ArkConfig as ArkBnConfig>::TWIST_TYPE;
+
+    const ATE_LOOP_COUNT: &'static [i8] = <ArkConfig as ArkBnConfig>::ATE_LOOP_COUNT;
+
+    // Possibly no longer useful here:
+    // const TWIST_MUL_BY_Q_X: Fq2
+    // const TWIST_MUL_BY_Q_Y: Fq2
```
Comment (lgiussan) [inline curves/bn254/src/curves/mod.rs]: Please remove if no longer necessary

File: models/src/models/bn/mod.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+    /// The absolute value of `6X + 2`.
+    const ATE_LOOP_COUNT: &'static [i8];
+
+    /// What kind of twist is this?
+    const TWIST_TYPE: TwistType;
+
+    // Redundant since mul_by_char is not used:
+    // const TWIST_MUL_BY_Q_X: Fp2<Self::Fp2Config>;
+    // const TWIST_MUL_BY_Q_Y: Fp2<Self::Fp2Config>;
```
Comment (lgiussan) [inline models/src/models/bn/mod.rs]: Please remove if no longer necessary

File: models/src/models/bn/g2.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+//     // multiply by field characteristic
+
+//     let mut s = r;
+//     s.x.frobenius_map_in_place(1);
+//     s.x *= &P::TWIST_MUL_BY_Q_X;
+//     s.y.frobenius_map_in_place(1);
+//     s.y *= &P::TWIST_MUL_BY_Q_Y;
+
+//     s
+// }
```
Comment (lgiussan) [inline models/src/models/bn/g2.rs]: Please remove if no longer necessary

File: LICENSE-APACHE [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+   To apply the Apache License to your work, attach the following
+   boilerplate notice, with the fields enclosed by brackets "[]"
+   replaced with your own identifying information. (Don't include
+   the brackets!)  The text should be enclosed in the appropriate
+   comment syntax for the file format. We also recommend that a
+   file or class name and description of purpose be included on the
+   same "printed page" as the copyright notice for easier
+   identification within third-party archives.
+
+Copyright 2022 arkworks contributors
```
Comment (lgiussan) [inline LICENSE-APACHE]: Please replace verbatim with
```
Copyright [yyyy] [name of copyright owner]
```

File: curves/bn254/src/curves/g1.rs [hunk 7]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2024, Horizen Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
```
Comment (lgiussan) [inline curves/bn254/src/curves/g1.rs]: If we intend to dual-license the software under MIT - Apache 2.0, I would propose to change all the copyright and license notices in the following way:
```
// Copyright 2022 arkworks contributors
// Copyright 2024 Horizen Labs, Inc.
// SPDX-License-Identifier: Apache-2.0 or MIT
```

File: Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+license = "MIT/Apache-2.0"
+edition = "2021"
+
+[workspace.dependencies]
+ark-ec = { version = "0.4.0", default-features = false }
+ark-ff = { version = "0.4.0", default-features = false }
+ark-std = { version = "0.4.0", default-features = false }
+ark-serialize = { version = "0.4.0", default-features = false }
+ark-bn254 = { version = "0.4.0", default-features = false, features = ["curve"] }
+ark-scale = { version = "0.0.12", default-features = false, features = ["hazmat"] }
```
Comment (lgiussan) [inline Cargo.toml:26]: Is the feature `hazmat` required?

File: test-utils/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+ark-scale.workspace = true
+
+[features]
+default = [ "std" ]
+std = [ "ark-ec/std", "ark-scale/std", "ark-std/std" ]
+# SCALE encoding related features for most of Arkworks types.
+# Defaults to "compressed" and "validated"
+scale-fast = [ "scale-no-compress", "scale-no-validate" ]
+scale-no-compress = []
+scale-no-validate = []
```
Comment (lgiussan) [inline test-utils/Cargo.toml]: The feature `scale-no-validate` doesn't seem to be used anywhere

File: Cargo.lock
Comment (lgiussan) [inline Cargo.lock:1]: Since this is a library, the `Cargo.lock` should not be committed

File: .gitignore [hunk 10]
```diff
@@ -0,0 +1,3 @@
+models/target/*
+target
+target/*
```
Comment (lgiussan) [inline .gitignore]: Please add the `Cargo.lock` file

**Review by lgiussan — CHANGES_REQUESTED** [review]
Nice work!
Just a few comments, mostly about formal stuff.

File: test-utils/Cargo.toml (hunk 9 above)
Comment (DimitrisPapac) [reply test-utils/Cargo.toml]: Only `scale-no-compress` is actually used on lines 36 and 38 of `test-utils/src/lib.rs`.

File: Cargo.toml (hunk 8 above)
Comment (DimitrisPapac) [reply Cargo.toml:26]: The code should compile and pass tests without the `hazmat` feature.

**Review by 95DDB — APPROVED** [review]

**Review by lgiussan — APPROVED** [review]

**Pull request merged**